hardware_interface/include/hardware_interface/actuator_interface.hpp

```diff
@@ -209,8 +230,34 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
   void set_state(const rclcpp_lifecycle::State & new_state) { lifecycle_state_ = new_state; }
 
 protected:
+  /// Get the logger of the ActuatorInterface.
+  /**
+   * \return logger of the ActuatorInterface.
+   */
+  rclcpp::Logger get_logger() const { return actuator_logger_; }
+
+  /// Get the clock of the ActuatorInterface.
+  /**
+   * \return clock of the ActuatorInterface.
+   */
+  rclcpp::Clock get_clock() const
+  {
+    if (clock_interface_)
+    {
+      return *(clock_interface_->get_clock());
+    }
+    else
+    {
+      return rclcpp::Clock(RCL_ROS_TIME);
+    }
+  }
```

Comment:
isn't this misleading? If we have no clock we shouldn't return any...

```suggestion
  std::optional<rclcpp::Clock> get_clock() const
  {
    return clock_interface_ ? std::optional<rclcpp::Clock>{*(clock_interface_->get_clock())} : std::nullopt;
  }
```

Reply:
Better yet, just use the existing sharedptr

```suggestion
  rclcpp::node_interfaces::NodeClockInterface::SharedPtr get_clock() const
  {
    return clock_interface_;
  }
```

Reply:
Sure, maybe shared_ptr is better and makes sense.

Reply:
Or maybe even better yet... can we validate the clock interface sharedptr we are getting and turn it into a clock? Is that dangerous?

Reply:
`*(clock_interface_->get_clock());` That's what I'm doing here right?

Reply:
Yes but what I mean is we should do this at the top and not even accept invalid clocks

Reply:
@bmagyar  Yes, I just pushed the changes!

With the new changes, we will need to push a fix to ros2_controllers as well inorder to fix some load tests. For instance this one

https://github.com/ros-controls/ros2_controllers/blob/b81be48fb1b6101a3c601c0ea11c5393e5a7026e/imu_sensor_broadcaster/test/test_load_imu_sensor_broadcaster.cpp#L34-L37

Before:
```
  controller_manager::ControllerManager cm(
    std::make_unique<hardware_interface::ResourceManager>(
      ros2_control_test_assets::minimal_robot_urdf),
    executor, "test_controller_manager");
```

Now:
```
controller_manager::ControllerManager cm(
    executor, ros2_control_test_assets::minimal_robot_urdf, "test_controller_manager");
```

Take a look at the changes and let me know if they look fine to you. If yes, I will open a PR for the controllers as well.

While making these modifications, I was under the dilemma to still have the `NodeClockInterface` and `NodeLoggingInterface` or just replace them with the `rclcpp::Clock::SharedPtr` and `rclcpp::Logger`. Just in case, I left the same.

Reply:
https://github.com/ros-controls/ros2_control/pull/1585

Reply:
Here is the PR of ros2_controllers: https://github.com/ros-controls/ros2_controllers/pull/1184

Reply:
Thank you!